pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -337,8 +339,83 @@ func (i *InfraProvider) DestroyBootstrap(dir string) error {
 	return nil
 }
 
+type machineManifest struct {
+	Status struct {
+		Addresses []clusterv1.MachineAddress `yaml:"addresses"`
+	} `yaml:"status"`
+}
+
+// extractIPAddress extracts the IP address from a machine manifest file in a
+// provider-agnostic way by reading only the "status" stanza, which should be
+// present in all providers.
+func extractIPAddress(manifestPath string) (string, error) {
+	data, err := os.ReadFile(manifestPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read machine manifest %s: %w", manifestPath, err)
+	}
+	var manifest machineManifest
+	if err := yaml.Unmarshal(data, &manifest); err != nil {
+		return "", fmt.Errorf("failed to unmarshal manifest %s: %w", manifestPath, err)
+	}
+
+	var ipAddr string
+	for _, addr := range manifest.Status.Addresses {
+		switch addr.Type {
+		case clusterv1.MachineExternalIP:
+			ipAddr = addr.Address
+		case clusterv1.MachineInternalIP:
+			// Prefer external IP when present
+			if len(ipAddr) == 0 {
+				ipAddr = addr.Address
+			}
+		default:
+			continue
+		}
+	}
+
+	return ipAddr, nil
+}
+
 // ExtractHostAddresses extracts the IPs of the bootstrap and control plane machines.
 func (i *InfraProvider) ExtractHostAddresses(dir string, config *types.InstallConfig, ha *infrastructure.HostAddresses) error {
+	logrus.Debugf("Looking for machine manifests in %s", dir)
+
+	bootstrapFiles, err := filepath.Glob(filepath.Join(dir, "*\\-openshift\\-cluster\\-api\\-guests\\-*\\-bootstrap.yaml"))
+	if err != nil {
+		return fmt.Errorf("failed to list bootstrap manifests: %w", err)
+	}
+	logrus.Debugf("bootstrap manifests found: %v", bootstrapFiles)
+
+	if len(bootstrapFiles) != 1 {
+		return fmt.Errorf("wrong number of bootstrap manifests found: %v. Expected exactly one", bootstrapFiles)
+	}
```

Comment:
Hm we have a slight problem here (a result of the fact that I only gave you a partial set of manifests as a basis for this). In CAPI, each machine has 2 manifests: a CAPI manifest and a platform-specific one. So I think we need a way to filter out the CAPI ones:

```
AWSClusterControllerIdentity--default.yaml                                      Machine-openshift-cluster-api-guests-padillon-03041822-fvhnc-bootstrap.yaml
AWSCluster-openshift-cluster-api-guests-padillon-03041822-fvhnc.yaml            Machine-openshift-cluster-api-guests-padillon-03041822-fvhnc-master-0.yaml
AWSMachine-openshift-cluster-api-guests-padillon-03041822-fvhnc-bootstrap.yaml  Machine-openshift-cluster-api-guests-padillon-03041822-fvhnc-master-1.yaml
AWSMachine-openshift-cluster-api-guests-padillon-03041822-fvhnc-master-0.yaml   Machine-openshift-cluster-api-guests-padillon-03041822-fvhnc-master-2.yaml
AWSMachine-openshift-cluster-api-guests-padillon-03041822-fvhnc-master-1.yaml   Namespace--openshift-cluster-api-guests.yaml
AWSMachine-openshift-cluster-api-guests-padillon-03041822-fvhnc-master-2.yaml   Secret-openshift-cluster-api-guests-padillon-03041822-fvhnc-bootstrap.yaml
Cluster-openshift-cluster-api-guests-padillon-03041822-fvhnc.yaml               Secret-openshift-cluster-api-guests-padillon-03041822-fvhnc-master.yaml

```

Reply:
> So I think we need a way to filter out the CAPI ones

Actually, it appears that both the CAPI & platform specific both have addresses, where CAPI's are copied from the machine provider:

https://github.com/kubernetes-sigs/cluster-api/blob/e6cc81a3a70a9e71cc1697bd31e78b5eea5b2847/api/v1beta1/machine_types.go#L194-L197

So it makes more sense to just directly use the CAPI one... So I guess the glob would look something like:

```go
"Machine\\-openshift\\-cluster\\-api\\-guests\\-*\\-bootstrap.yaml"
```

Reply:
Done.